File: src/ApiSpec.php
<?php

declare(strict_types=1);

namespace XstrMe\ApiSpec;

use Symfony\Component\Yaml\Yaml;
use InvalidArgumentException;
use RuntimeException;

/**
 * Utility class for accessing the XStr.me API specification in PHP applications.
 * 
 * This class provides convenient methods to load and access the OpenAPI
 * specification that is bundled with this package.
 * 
 * @package XstrMe\ApiSpec
 * @version 1.0.0
 */
final class ApiSpec
{
    private const API_SPEC_PATH = __DIR__ . '/../api-spec.yml';
    
    private static ?array $cachedSpec = null;
    private static ?string $cachedYaml = null;
    
    /**
     * Get the API specification as a PHP array.
     * 
     * @return array The parsed API specification
     * @throws RuntimeException if the API specification cannot be found or parsed
     */
    public static function getApiSpecAsArray(): array
    {
        if (self::$cachedSpec === null) {
            self::$cachedSpec = self::loadAndParseSpec();
        }
        
        return self::$cachedSpec;
    }
    
    /**
     * Get the API specification as YAML string.
     * 
     * @return string The API specification YAML content
     * @throws RuntimeException if the API specification cannot be found
     */
    public static function getApiSpecAsYaml(): string
    {
        if (self::$cachedYaml === null) {
            if (!file_exists(self::API_SPEC_PATH)) {
                throw new RuntimeException(
                    sprintf('API specification not found at: %s', self::API_SPEC_PATH)
                );
            }
            
            $content = file_get_contents(self::API_SPEC_PATH);
            if ($content === false) {
                throw new RuntimeException(
                    sprintf('Failed to read API specification from: %s', self::API_SPEC_PATH)
                );
            }
            
            self::$cachedYaml = $content;
        }
        
        return self::$cachedYaml;
    }
    
    /**
     * Get the API specification as JSON string.
     * 
     * @param int $flags JSON encoding flags
     * @return string The API specification as JSON
     * @throws RuntimeException if the API specification cannot be converted to JSON
     */
    public static function getApiSpecAsJson(int $flags = JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES): string
    {
        $spec = self::getApiSpecAsArray();
        
        $json = json_encode($spec, $flags);
        if ($json === false) {
            throw new RuntimeException('Failed to encode API specification as JSON: ' . json_last_error_msg());
        }
        
        return $json;
    }
    
    /**
     * Get the version of the API specification.
     * 
     * @return string API specification version
     */
    public static function getVersion(): string
    {
        $spec = self::getApiSpecAsArray();
        return $spec['info']['version'] ?? '1.0.0';
    }
    
    /**
     * Get the title of the API specification.
     * 
     * @return string API specification title
     */
    public static function getTitle(): string
    {
        $spec = self::getApiSpecAsArray();
        return $spec['info']['title'] ?? 'XStr.me API';
    }
    
    /**
     * Get all available paths from the API specification.
     * 
     * @return array List of API paths
     */
    public static function getPaths(): array
    {
        $spec = self::getApiSpecAsArray();
        return array_keys($spec['paths'] ?? []);
    }
    
    /**
     * Get all available servers from the API specification.
     * 
     * @return array List of API servers
     */
    public static function getServers(): array
    {
        $spec = self::getApiSpecAsArray();
        return $spec['servers'] ?? [];
    }
    
    /**
     * Check if the API specification file is available.
     * 
     * @return bool true if the API specification is available, false otherwise
     */
    public static function isAvailable(): bool
    {
        return file_exists(self::API_SPEC_PATH) && is_readable(self::API_SPEC_PATH);
    }
    
    /**
     * Get a specific path definition from the API specification.
     * 
     * @param string $path The API path (e.g., '/health')
     * @return array|null The path definition or null if not found
     */
    public static function getPathDefinition(string $path): ?array
    {
        $spec = self::getApiSpecAsArray();
        return $spec['paths'][$path] ?? null;
    }
    
    /**
     * Get all schema definitions from the API specification.
     * 
     * @return array All schema definitions
     */
    public static function getSchemas(): array
    {
        $spec = self::getApiSpecAsArray();
        return $spec['components']['schemas'] ?? [];
    }
    
    /**
     * Get a specific schema definition.
     * 
     * @param string $schemaName The name of the schema
     * @return array|null The schema definition or null if not found
     */
    public static function getSchema(string $schemaName): ?array
    {
        $schemas = self::getSchemas();
        return $schemas[$schemaName] ?? null;
    }
    
    /**
     * Clear the internal cache.
     * 
     * @return void
     */
    public static function clearCache(): void
    {
        self::$cachedSpec = null;
        self::$cachedYaml = null;
    }
    
    /**
     * Load and parse the API specification file.
     * 
     * @return array The parsed API specification
     * @throws RuntimeException if the file cannot be found or parsed
     */
    private static function loadAndParseSpec(): array
    {
        if (!file_exists(self::API_SPEC_PATH)) {
            throw new RuntimeException(
                sprintf('API specification not found at: %s', self::API_SPEC_PATH)
            );
        }
        
        try {
            $content = file_get_contents(self::API_SPEC_PATH);
            if ($content === false) {
                throw new RuntimeException(
                    sprintf('Failed to read API specification from: %s', self::API_SPEC_PATH)
                );
            }
            
            $parsed = Yaml::parse($content);
            if (!is_array($parsed)) {
                throw new RuntimeException('Invalid YAML content in API specification');
            }
            
            return $parsed;
        } catch (\Exception $e) {
            throw new RuntimeException(
                sprintf('Failed to parse API specification: %s', $e->getMessage()),
                0,
                $e
            );
        }
    }
}

File: src/python/xstr_me_api_spec/api_spec.py
"""
Utility class for accessing the XStr.me API specification in Python applications.

This module provides convenient methods to load and access the OpenAPI
specification that is bundled with this package.
"""

import json
import os
from pathlib import Path
from typing import Dict, Any, Optional

try:
    import yaml
except ImportError:
    yaml = None


class ApiSpecError(Exception):
    """Base exception for API specification related errors."""
    pass


class ApiSpec:
    """
    Utility class for accessing the XStr.me API specification.
    
    This class provides convenient methods to load and access the OpenAPI
    specification that is bundled with this package.
    
    Example:
        >>> from xstr_me_api_spec import ApiSpec
        >>> spec_dict = ApiSpec.get_api_spec_as_dict()
        >>> spec_yaml = ApiSpec.get_api_spec_as_yaml()
        >>> spec_json = ApiSpec.get_api_spec_as_json()
    """
    
    _API_SPEC_PATH = Path(__file__).parent / "api-spec.yml"
    _cached_spec: Optional[Dict[str, Any]] = None
    _cached_yaml: Optional[str] = None
    
    def __init__(self) -> None:
        """Private constructor to prevent instantiation."""
        raise AssertionError("Utility class should not be instantiated")
    
    @classmethod
    def get_api_spec_as_dict(cls) -> Dict[str, Any]:
        """
        Get the API specification as a Python dictionary.
        
        Returns:
            Dict[str, Any]: The parsed API specification
            
        Raises:
            ApiSpecError: If the API specification cannot be found or parsed
            ImportError: If PyYAML is not installed
        """
        if cls._cached_spec is None:
            cls._cached_spec = cls._load_and_parse_spec()
        
        return cls._cached_spec.copy()
    
    @classmethod
    def get_api_spec_as_yaml(cls) -> str:
        """
        Get the API specification as YAML string.
        
        Returns:
            str: The API specification YAML content
            
        Raises:
            ApiSpecError: If the API specification cannot be found
        """
        if cls._cached_yaml is None:
            if not cls._API_SPEC_PATH.exists():
                raise ApiSpecError(
                    f"API specification not found at: {cls._API_SPEC_PATH}"
                )
            
            try:
                cls._cached_yaml = cls._API_SPEC_PATH.read_text(encoding='utf-8')
            except OSError as e:
                raise ApiSpecError(
                    f"Failed to read API specification: {e}"
                ) from e
        
        return cls._cached_yaml
    
    @classmethod
    def get_api_spec_as_json(cls, indent: Optional[int] = None) -> str:
        """
        Get the API specification as JSON string.
        
        Args:
            indent: Number of spaces for JSON indentation. None for compact format.
        
        Returns:
            str: The API specification JSON content
            
        Raises:
            ApiSpecError: If the API specification cannot be found or parsed
        """
        spec_dict = cls.get_api_spec_as_dict()
        try:
            return json.dumps(spec_dict, indent=indent, ensure_ascii=False)
        except (TypeError, ValueError) as e:
            raise ApiSpecError(
                f"Failed to serialize API specification to JSON: {e}"
            ) from e
    
    @classmethod
    def get_api_version(cls) -> str:
        """
        Get the API version from the specification.
        
        Returns:
            str: The API version
            
        Raises:
            ApiSpecError: If the API specification cannot be found or parsed
        """
        spec = cls.get_api_spec_as_dict()
        try:
            return spec["info"]["version"]
        except KeyError as e:
            raise ApiSpecError(
                f"API version not found in specification: {e}"
            ) from e
    
    @classmethod
    def get_api_title(cls) -> str:
        """
        Get the API title from the specification.
        
        Returns:
            str: The API title
            
        Raises:
            ApiSpecError: If the API specification cannot be found or parsed
        """
        spec = cls.get_api_spec_as_dict()
        try:
            return spec["info"]["title"]
        except KeyError as e:
            raise ApiSpecError(
                f"API title not found in specification: {e}"
            ) from e
    
    @classmethod
    def get_api_description(cls) -> str:
        """
        Get the API description from the specification.
        
        Returns:
            str: The API description
            
        Raises:
            ApiSpecError: If the API specification cannot be found or parsed
        """
        spec = cls.get_api_spec_as_dict()
        try:
            return spec["info"]["description"]
        except KeyError as e:
            raise ApiSpecError(
                f"API description not found in specification: {e}"
            ) from e
    
    @classmethod
    def get_servers(cls) -> list:
        """
        Get the server configurations from the specification.
        
        Returns:
            list: List of server configurations
            
        Raises:
            ApiSpecError: If the API specification cannot be found or parsed
        """
        spec = cls.get_api_spec_as_dict()
        return spec.get("servers", [])
    
    @classmethod
    def get_paths(cls) -> Dict[str, Any]:
        """
        Get the API paths from the specification.
        
        Returns:
            Dict[str, Any]: Dictionary of API paths and their configurations
            
        Raises:
            ApiSpecError: If the API specification cannot be found or parsed
        """
        spec = cls.get_api_spec_as_dict()
        return spec.get("paths", {})
    
    @classmethod
    def _load_and_parse_spec(cls) -> Dict[str, Any]:
        """
        Load and parse the API specification file.
        
        Returns:
            Dict[str, Any]: The parsed API specification
            
        Raises:
            ApiSpecError: If the API specification cannot be found or parsed
            ImportError: If PyYAML is not installed
        """
        if yaml is None:
            raise ImportError(
                "PyYAML is required to parse the API specification. "
                "Install it with: pip install pyyaml"
            )
        
        if not cls._API_SPEC_PATH.exists():
            raise ApiSpecError(
                f"API specification not found at: {cls._API_SPEC_PATH}"
            )
        
        try:
            yaml_content = cls._API_SPEC_PATH.read_text(encoding='utf-8')
            return yaml.safe_load(yaml_content)
        except OSError as e:
            raise ApiSpecError(
                f"Failed to read API specification: {e}"
            ) from e
        except yaml.YAMLError as e:
            raise ApiSpecError(
                f"Failed to parse API specification YAML: {e}"
            ) from e
    
    @classmethod
    def clear_cache(cls) -> None:
        """
        Clear the internal cache for the API specification.
        
        This method is useful for testing or when the specification file
        might have changed during runtime.
        """
        cls._cached_spec = None
        cls._cached_yaml = None
